File: LV2/Z2.py
from machine import Pin
from time import sleep

T1 = Pin(0, Pin.IN)  # Increment
T2 = Pin(1, Pin.IN)  # Decrement
T3 = Pin(2, Pin.IN)  # Set sum to 0x00
T4 = Pin(3, Pin.IN)  # Set sum to 0xFF

# Binary digits, 8-bit number with digits: 76543210
LED0 = Pin(12, Pin.OUT)
LED1 = Pin(13, Pin.OUT)
LED2 = Pin(14, Pin.OUT)
LED3 = Pin(15, Pin.OUT)
LED4 = Pin(4, Pin.OUT)
LED5 = Pin(5, Pin.OUT)
LED6 = Pin(6, Pin.OUT)
LED7 = Pin(7, Pin.OUT)

outputLEDs = [LED0, LED1, LED2, LED3, LED4, LED5, LED6, LED7]

button1 = Pin(0, Pin.IN)


def setZeroes():
    for pin in outputLEDs:
        pin.value(0)


def setOnes():
    for pin in outputLEDs:
        pin.value(1)


def increment():
    carry = 1

    for pin in outputLEDs:
        current = pin.value()

        if current == 1 and carry == 1:
            pin.value(0)
        else:
            new = current + carry
            pin.value(new)
            carry = 0


def decrement():
    index = 0
    
    while index < len(outputLEDs):
        if outputLEDs[index].value() == 1:
            break
        else:
            index = index + 1

    if index != len(outputLEDs):
        outputLEDs[index].value(0)

    for i in range(index):
        outputLEDs[i].value(1)
        

def main():
    setOnes()
    sleep(1)
    setZeroes()

    while True:

        sleep(0.3)
        while True:
            sleep(0.3)
            increment()

            if button1.value():

                break
        
        sleep(0.3)
        while True:
            decrement()
            sleep(0.3)

            if button1.value():
                break

        
            

if __name__ == "__main__":
    main()

File: LV2/Z2.cpp
#include "mbed.h"
#include "lpc1114etf.h"

DigitalOut E(LED_ACT);

DigitalOut led1(LED0), led2(LED1), led3(LED2), led4(LED3), led5(LED4), led6(LED5), led7(LED6), led8(LED7);

DigitalOut LEDs[] = {led1, led2, led3, led4, led5, led6, led7, led8};

DigitalIn Btn(Taster_1);

bool countUp = true;

void increment() {
    int carry = 1;

    for (auto &pin : LEDs) {
        int current = pin;
        if (current == 1 && carry == 1) {
            pin = 0;
        } else {
            int tmp = carry + current;
            pin = tmp;
            carry = 0;
        }
    }
}

void decrement() {
    int index = 0;

    while (index < 8) {
        if (LEDs[index] == 1) {
            break;
        } else {
            index++;
        }
    }

    if (index != 8) {
        LEDs[index] = 0;
    }

    for (int i = 0; i < index; i++) {
        LEDs[i] = 1;
    }
}

void changeDirection() {
    countUp = !countUp;
}

int main() {
    E = 0;

        while (true) {
            if (countUp)
                increment();
            else
                decrement();

            wait_us(1000000);

            if (Btn == 1)
                changeDirection();
        }
    
}